stream.go

```diff
@@ -0,0 +1,451 @@
+
+// String returns the ID in the format <time>-<seq> (the same format used by Redis).
+//
+// String implements the fmt.Stringer interface.
+func (s StreamEntryID) String() string {
+	var buf [maxUint64Len*2 + 1]byte
+	b := strconv.AppendUint(buf[:0], s.Time, 10)
+	b = append(b, '-')
+	b = strconv.AppendUint(b, s.Seq, 10)
+	return string(b)
+}
+
+// StreamEntry is an entry in a Redis stream as returned by XRANGE, XREAD and XREADGROUP.
+type StreamEntry struct {
+	// ID is the ID of the entry in a stream.
+	ID StreamEntryID
+
+	// Fields contains the fields and values for the stream entry.
+	Fields map[string]string
+}
+
+var _ resp.Unmarshaler = (*StreamEntry)(nil)
+
+var errInvalidStreamEntry = errors.New("invalid stream entry")
+
+// UnmarshalRESP implements the resp.Unmarshaler interface.
+func (s *StreamEntry) UnmarshalRESP(br *bufio.Reader) error {
+	var ah resp2.ArrayHeader
+	if err := ah.UnmarshalRESP(br); err != nil {
+		return err
+	}
+	if ah.N != 2 {
+		return errInvalidStreamEntry
+	}
+
+	if err := s.ID.UnmarshalRESP(br); err != nil {
+		return err
+	}
+
+	return s.unmarshalFields(br)
+}
+
+func (s *StreamEntry) unmarshalFields(br *bufio.Reader) error {
+	// resp2.Any{I: &s.Fields}.UnmarshalRESP(br)
+	var ah resp2.ArrayHeader
+	if err := ah.UnmarshalRESP(br); err != nil {
+		return err
+	}
+	if ah.N%2 != 0 {
+		return errInvalidStreamEntry
+	}
+
+	if s.Fields == nil {
+		s.Fields = make(map[string]string, ah.N/2)
+	} else {
+		for k := range s.Fields {
+			delete(s.Fields, k)
+		}
+	}
+
+	var bs resp2.BulkString
+	for i := 0; i < ah.N; i += 2 {
+		if err := bs.UnmarshalRESP(br); err != nil {
+			return err
+		}
+		key := bs.S
+		if err := bs.UnmarshalRESP(br); err != nil {
+			return err
+		}
+		s.Fields[key] = bs.S
+	}
+	return nil
+}
+
+// StreamReaderOpts contains various options given for NewStreamReader that influence the behaviour.
+//
+// The only required field is Streams.
+type StreamReaderOpts struct {
+	// Streams must contain one or more stream names that will be read.
+	//
+	// The value for each stream can either be nil or an existing ID.
+	// If a value is non-nil, only newer stream entries will be returned.
+	Streams map[string]*StreamEntryID
+
+	// Group is an optional consumer group name.
+	//
+	// If Group is not empty reads will use XREADGROUP with the Group as consumer group instead of XREAD.
+	Group string
+
+	// Consumer is an optional consumer name for use with Group.
+	//
+	// TODO(nussjustin): Automatically detect if empty and Group is set?
+	Consumer string
+
+	// Noack optionally enables passing the NOACK flag to XREADGROUP.
+	Noack bool
+
+	// Block specifies the duration in milliseconds that reads will wait for new data before returning.
+	//
+	// If Block is negative, reads will not block (no BLOCK option is set) when there is no new data to read.
+	//
+	// TODO(nussjustin): Make no blocking mode default? How to specify 0 milliseconds (infinite blocking)?
+	Block time.Duration
```

Comment:
Just an idea, but most users will probably want to wait indefinitely, so the current behaviour seems best to me.

Reply:
We would still need a way to specify infinite blocking and no blocking. We could change this to a pointer, but that would be annoying for users. Also most (all?) functions don't handle dropped connections, etc. if I'm not mistaken (if they did, then the StreamReader would do to automatically). Maybe the default Client just needs a keep alive for TCP connections?

Reply:
> We would still need a way to specify infinite blocking and no blocking.

How about adding another field, `NoBlock bool`. Then if `Block` is negative that means infinite, and zero means the default.

>  Also most (all?) functions don't handle dropped connections

The connections don't handle keepalive on their own, but everything in radix which uses them implicitly does. The Pool sends PINGs periodically, Sentinel calls `SENTINEL SENTINELS` periodically on its connection to the sentinel itself, etc... Basically all connections have data sent over them periodically to keep them from getting stale and to ensure the server is still responsive. If they didn't then they'd be at the mercy of the OS to determine that a connection is closed, which can take minutes sometimes.

> Maybe the default Client just needs a keep alive for TCP connections?

That wouldn't handle the case where the server becomes unresponsive, unfortunately. TCP keepalive is useful when your endpoint doesn't have any sort of application-layer pinging mechanism, but it's better to use an app-layer one if possible.

Reply:
The problem is that on a connection level there are no timeouts. Once I have a connection (either from Dial or from a Pool) the connection itself will not timeout and hang if there is a connection problem that does not close the socket.

PINGing the server from another connection doesn't help here, since the hanging connection will still hang even if the Pool notices that there is a problem.

For users that care about this, there is currently no direct way to work around this in radix.

Since radix does not know if a command is blocking, setting a timeout automatically is also not possible and also setting a timeout for each command would involve some overhead.

TCP keepalive doesn't completely solve this, but could at least reduce the impact of hanging connections.

> How about adding another field, NoBlock bool. Then if Block is negative that means infinite, and zero means the default.

Done